File: packages/js/cli/src/templates/typescript.ts
/**
 * v2 TypeScript scaffold template.
 *
 * Generates a complete v2 trik project structure using the
 * agent-based handoff architecture with wrapAgent() from @trikhub/sdk.
 */

// ============================================================================
// Types
// ============================================================================

export interface InitConfig {
  name: string;
  displayName: string;
  description: string;
  authorName: string;
  authorGithub: string;
  category: string;
  enableStorage: boolean;
  enableConfig: boolean;
  // v2 fields
  agentMode: 'conversational' | 'tool';
  handoffDescription: string;
  domainTags: string[];
  toolNames: string[];
}

// ============================================================================
// File generators
// ============================================================================

function generateManifest(config: InitConfig): string {
  const isToolMode = config.agentMode === 'tool';

  const agent: Record<string, unknown> = {
    mode: config.agentMode,
    domain: config.domainTags,
  };

  if (!isToolMode) {
    agent.handoffDescription = config.handoffDescription;
    agent.systemPromptFile = './src/prompts/system.md';
    agent.model = { capabilities: ['tool_use'] };
  }

  const manifest: Record<string, unknown> = {
    schemaVersion: 2,
    id: config.name,
    name: config.displayName,
    description: config.description,
    version: '0.1.0',
    agent,
  };

  // Tools block
  if (isToolMode && config.toolNames.length > 0) {
    const tools: Record<string, Record<string, unknown>> = {};
    for (const toolName of config.toolNames) {
      tools[toolName] = {
        description: `TODO: describe ${toolName}`,
        inputSchema: {
          type: 'object',
          properties: {
            query: { type: 'string', maxLength: 200 },
          },
          required: ['query'],
        },
        outputSchema: {
          type: 'object',
          properties: {
            result: { type: 'string', maxLength: 500 },
          },
          required: ['result'],
        },
      };
    }
    manifest.tools = tools;
  } else {
    manifest.tools = {
      exampleTool: {
        description: 'An example tool',
      },
    };
  }

  if (config.enableStorage) {
    manifest.capabilities = {
      storage: { enabled: true },
    };
  }

  manifest.limits = { maxTurnTimeMs: 30000 };
  manifest.entry = {
    module: './dist/agent.js',
    export: 'default',
  };
  manifest.author = config.authorName;

  if (config.enableConfig) {
    manifest.config = {
      optional: [
        { key: 'ANTHROPIC_API_KEY', description: 'Anthropic API key for the agent' },
      ],
    };
  }

  return JSON.stringify(manifest, null, 2);
}

function generateTrikhubJson(config: InitConfig): string {
  const metadata = {
    displayName: config.displayName,
    shortDescription: config.description,
    categories: [config.category],
    keywords: [] as string[],
    author: {
      name: config.authorName,
      github: config.authorGithub,
    },
    repository: `https://github.com/${config.authorGithub}/${config.name}`,
  };

  return JSON.stringify(metadata, null, 2);
}

function generatePackageJson(config: InitConfig): string {
  const isToolMode = config.agentMode === 'tool';

  const dependencies: Record<string, string> = {
    '@trikhub/sdk': 'latest',
  };

  // Conversational mode needs LangChain + LLM provider
  if (!isToolMode) {
    dependencies['@langchain/anthropic'] = '^0.3.0';
    dependencies['@langchain/core'] = '^0.3.0';
    dependencies['@langchain/langgraph'] = '^0.2.0';
    dependencies['zod'] = '^3.25.0';
  }

  const pkg = {
    name: config.name,
    version: '0.1.0',
    description: config.description,
    type: 'module',
    main: './dist/agent.js',
    scripts: {
      build: 'tsc',
      dev: 'node --import tsx src/agent.ts',
      clean: 'rm -rf dist *.tsbuildinfo',
    },
    dependencies,
    devDependencies: {
      tsx: '^4.19.0',
      typescript: '^5.7.0',
    },
  };

  return JSON.stringify(pkg, null, 2);
}

function generateTsConfig(): string {
  const tsconfig = {
    compilerOptions: {
      target: 'ES2022',
      module: 'NodeNext',
      moduleResolution: 'nodenext',
      outDir: './dist',
      rootDir: './src',
      strict: true,
      esModuleInterop: true,
      skipLibCheck: true,
      forceConsistentCasingInFileNames: true,
      declaration: true,
      declarationMap: true,
      sourceMap: true,
    },
    include: ['src/**/*'],
  };

  return JSON.stringify(tsconfig, null, 2);
}

function generateToolModeAgentTs(config: InitConfig): string {
  const handlers = config.toolNames.map((name) =>
    `  ${name}: async (input, context) => {
    // TODO: Implement ${name}
    return { result: 'Not implemented' };
  },`
  ).join('\n');

  return `/**
 * ${config.displayName} — tool-mode agent entry point.
 *
 * Exports native tools to the main agent. No handoff, no session.
 */

import { wrapToolHandlers } from '@trikhub/sdk';

export default wrapToolHandlers({
${handlers}
});
`;
}

function generateAgentTs(config: InitConfig): string {
  if (config.agentMode === 'tool') {
    return generateToolModeAgentTs(config);
  }

  // Conversational mode
  return `/**
 * ${config.displayName} — conversational agent entry point.
 */

import { ChatAnthropic } from '@langchain/anthropic';
import { createReactAgent } from '@langchain/langgraph/prebuilt';
import { wrapAgent, transferBackTool } from '@trikhub/sdk';
import type { TrikContext } from '@trikhub/sdk';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { exampleTool } from './tools/example.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const systemPrompt = readFileSync(join(__dirname, '../src/prompts/system.md'), 'utf-8');

export default wrapAgent((context: TrikContext) => {
  const model = new ChatAnthropic({
    modelName: 'claude-sonnet-4-20250514',
    anthropicApiKey: ${config.enableConfig ? "config.get('ANTHROPIC_API_KEY')" : 'process.env.ANTHROPIC_API_KEY'},
  });

  const tools = [
    exampleTool,
    transferBackTool,
  ];

  return createReactAgent({
    llm: model,
    tools,
    messageModifier: systemPrompt,
  });
});
`;
}

function generateExampleTool(): string {
  return `import { tool } from '@langchain/core/tools';
import { z } from 'zod';

export const exampleTool = tool(
  async (input) => {
    // TODO: Implement your tool logic here
    return JSON.stringify({ result: \`Processed: \${input.query}\` });
  },
  {
    name: 'exampleTool',
    description: 'An example tool — replace with your own implementation',
    schema: z.object({
      query: z.string().describe('The input query to process'),
    }),
  },
);
`;
}

function generateSystemPrompt(config: InitConfig): string {
  const domainStr = config.domainTags.join(', ');

  return `# ${config.displayName}

You are ${config.displayName}, a specialized assistant for ${config.description.toLowerCase()}.

## Your capabilities
- **exampleTool**: An example tool

## Guidelines
- Focus on tasks within your domain: ${domainStr}
- When the user's request is outside your expertise, use the transfer_back tool
- Provide clear, actionable responses

## Transfer back
Use the \`transfer_back\` tool when:
- The user's request is outside your domain
- You've completed the task and the user wants to do something else
- The user explicitly asks to go back
`;
}

function generateGitignore(): string {
  return `node_modules/
dist/
*.tsbuildinfo
.env
.trikhub/secrets.json
`;
}

function generateReadme(config: InitConfig): string {
  const domainStr = config.domainTags.join(', ');
  const isToolMode = config.agentMode === 'tool';

  const devSection = isToolMode
    ? `- Implement your tool handlers in \`src/agent.ts\`
- Update inputSchema/outputSchema in \`manifest.json\``
    : `- Edit your agent logic in \`src/agent.ts\`
- Add tools in \`src/tools/\`
- Customize the system prompt in \`src/prompts/system.md\``;

  const archSection = isToolMode
    ? `Tools from this trik appear as native tools on the main agent — no handoff, no session.`
    : `The main agent routes conversations to this trik using a \`talk_to_${config.name}\` tool.
When done, use the \`transfer_back\` tool to return control.`;

  return `# ${config.displayName}

${config.description}

## Getting Started

1. Install dependencies: \`npm install\`
2. Build: \`npm run build\`
3. Validate: \`trik lint .\`
4. Publish: \`trik publish\`

## Development

${devSection}

## Architecture

This trik uses the TrikHub v2 architecture:
- **Mode**: ${config.agentMode}
- **Domain**: ${domainStr}

${archSection}
`;
}

// ============================================================================
// Public API
// ============================================================================

/**
 * Generate a complete v2 TypeScript trik project.
 *
 * @returns Map of { relativePath: fileContent } for all project files.
 */
export function generateTypescriptProject(config: InitConfig): Record<string, string> {
  const files: Record<string, string> = {};

  // Core config files
  files['manifest.json'] = generateManifest(config);
  files['trikhub.json'] = generateTrikhubJson(config);
  files['package.json'] = generatePackageJson(config);
  files['tsconfig.json'] = generateTsConfig();
  files['.gitignore'] = generateGitignore();
  files['README.md'] = generateReadme(config);

  // Source files
  files['src/agent.ts'] = generateAgentTs(config);

  if (config.agentMode === 'conversational') {
    // Conversational mode: example tool + system prompt
    files['src/tools/example.ts'] = generateExampleTool();
    files['src/prompts/system.md'] = generateSystemPrompt(config);
  }
  // Tool mode: no separate tool files or system prompt — all logic in agent.ts

  return files;
}

File: packages/python/trikhub-cli/src/trikhub/cli/templates/typescript.py
"""TypeScript Trik Template Generator (v2).

Generates all files needed for a v2 TypeScript trik project.
Supports conversational and tool modes using the @trikhub/sdk.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any


@dataclass
class TsTemplateConfig:
    name: str
    display_name: str
    description: str
    author_name: str
    author_github: str
    category: str
    enable_storage: bool
    enable_config: bool
    agent_mode: str  # 'conversational' | 'tool'
    handoff_description: str = ""
    domain_tags: list[str] = field(default_factory=list)
    tool_names: list[str] = field(default_factory=list)


@dataclass
class GeneratedFile:
    path: str
    content: str


def generate_typescript_project(config: TsTemplateConfig) -> list[GeneratedFile]:
    files: list[GeneratedFile] = []

    files.append(GeneratedFile("manifest.json", _generate_manifest(config)))
    files.append(GeneratedFile("trikhub.json", _generate_trikhub_json(config)))
    files.append(GeneratedFile("package.json", _generate_package_json(config)))
    files.append(GeneratedFile("tsconfig.json", _generate_tsconfig()))
    files.append(GeneratedFile(".gitignore", _generate_gitignore()))

    if config.agent_mode == "conversational":
        files.append(GeneratedFile("src/agent.ts", _generate_conversational_agent(config)))
        files.append(GeneratedFile("src/tools/example.ts", _generate_example_tool_ts()))
        files.append(GeneratedFile("src/prompts/system.md", _generate_system_prompt(config)))
    else:
        files.append(GeneratedFile("src/agent.ts", _generate_tool_agent(config)))

    return files


def _generate_manifest(config: TsTemplateConfig) -> str:
    agent: dict[str, Any] = {
        "mode": config.agent_mode,
        "domain": config.domain_tags or [config.category],
    }

    if config.agent_mode == "conversational":
        agent["handoffDescription"] = config.handoff_description or config.description
        agent["systemPromptFile"] = "./src/prompts/system.md"
        agent["model"] = {"capabilities": ["tool_use"]}

    manifest: dict[str, Any] = {
        "schemaVersion": 2,
        "id": config.name,
        "name": config.display_name,
        "description": config.description,
        "version": "0.1.0",
        "agent": agent,
        "limits": {"maxTurnTimeMs": 30000},
        "entry": {
            "module": "./dist/agent.js",
            "export": "default",
            "runtime": "node",
        },
        "author": f"@{config.author_github}",
        "repository": f"https://github.com/{config.author_github}/{config.name}",
    }

    if config.agent_mode == "tool":
        tools: dict[str, Any] = {}
        for tool_name in (config.tool_names or ["exampleTool"]):
            tools[tool_name] = {
                "description": f"TODO: describe {tool_name}",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "query": {"type": "string", "description": "Input query"},
                    },
                    "required": ["query"],
                },
                "outputSchema": {
                    "type": "object",
                    "properties": {
                        "result": {
                            "type": "string",
                            "maxLength": 1000,
                            "pattern": "^.{0,1000}$",
                        },
                    },
                },
                "outputTemplate": "Result: {{result}}",
            }
        manifest["tools"] = tools
    else:
        manifest["tools"] = {
            "exampleTool": {
                "description": "An example tool",
            },
        }

    if config.enable_storage:
        manifest["capabilities"] = {
            "storage": {"enabled": True, "maxSizeBytes": 1048576, "persistent": True},
        }

    if config.enable_config:
        manifest["config"] = {
            "required": [{"key": "API_KEY", "description": "Your API key"}],
            "optional": [],
        }

    return json.dumps(manifest, indent=2) + "\n"


def _generate_trikhub_json(config: TsTemplateConfig) -> str:
    trikhub = {
        "displayName": config.display_name,
        "shortDescription": config.description,
        "categories": [config.category],
        "keywords": [config.name],
        "author": {
            "name": config.author_name,
            "github": config.author_github,
        },
        "repository": f"https://github.com/{config.author_github}/{config.name}",
    }
    return json.dumps(trikhub, indent=2) + "\n"


def _generate_package_json(config: TsTemplateConfig) -> str:
    deps: dict[str, str] = {"@trikhub/sdk": "^0.8.0"}
    if config.agent_mode == "conversational":
        deps.update({
            "@langchain/anthropic": "^0.3.0",
            "@langchain/core": "^0.3.0",
            "@langchain/langgraph": "^0.2.0",
            "zod": "^3.22.0",
        })

    pkg = {
        "name": f"@{config.author_github.lower()}/{config.name}",
        "version": "0.1.0",
        "description": config.description,
        "type": "module",
        "main": "dist/agent.js",
        "scripts": {
            "build": "tsc",
            "clean": "rm -rf dist",
        },
        "dependencies": deps,
        "devDependencies": {
            "@types/node": "^20.0.0",
            "typescript": "^5.6.0",
        },
        "engines": {"node": ">=20"},
    }
    return json.dumps(pkg, indent=2) + "\n"


def _generate_tsconfig() -> str:
    tsconfig = {
        "compilerOptions": {
            "target": "ES2022",
            "module": "NodeNext",
            "moduleResolution": "NodeNext",
            "outDir": "./dist",
            "rootDir": "./src",
            "strict": True,
            "esModuleInterop": True,
            "skipLibCheck": True,
            "declaration": True,
        },
        "include": ["src/**/*"],
        "exclude": ["node_modules", "dist"],
    }
    return json.dumps(tsconfig, indent=2) + "\n"


def _generate_conversational_agent(config: TsTemplateConfig) -> str:
    return f'''/**
 * {config.display_name} — conversational agent
 *
 * Uses LangGraph ReAct pattern with the TrikHub SDK.
 */

import {{ ChatAnthropic }} from '@langchain/anthropic';
import {{ createReactAgent }} from '@langchain/langgraph/prebuilt';
import {{ wrapAgent, transferBackTool }} from '@trikhub/sdk';
import type {{ TrikContext }} from '@trikhub/sdk';

import {{ exampleTool }} from './tools/example.js';

export default wrapAgent((context: TrikContext) => {{
  const model = new ChatAnthropic({{ modelName: 'claude-sonnet-4-20250514' }});
  const tools = [exampleTool, transferBackTool];

  return createReactAgent({{
    llm: model,
    tools,
  }});
}});
'''


def _generate_tool_agent(config: TsTemplateConfig) -> str:
    handlers: list[str] = []
    for tool_name in (config.tool_names or ["exampleTool"]):
        handlers.append(f"""  {tool_name}: async (input, context) => {{
    const query = (input as {{ query: string }}).query;
    return {{ result: `Processed: ${{query}}` }};
  }},""")

    handlers_str = "\n".join(handlers)
    return f'''/**
 * {config.display_name} — tool mode agent
 *
 * Exports native tools via wrapToolHandlers from the TrikHub SDK.
 */

import {{ wrapToolHandlers }} from '@trikhub/sdk';

export default wrapToolHandlers({{
{handlers_str}
}});
'''


def _generate_example_tool_ts() -> str:
    return '''/**
 * Example tool for the conversational agent.
 */

import { tool } from '@langchain/core/tools';
import { z } from 'zod';

export const exampleTool = tool(
  async ({ query }) => {
    return `Result for: ${query}`;
  },
  {
    name: 'exampleTool',
    description: 'Search for information about a topic',
    schema: z.object({
      query: z.string().describe('The search query'),
    }),
  },
);
'''


def _generate_system_prompt(config: TsTemplateConfig) -> str:
    return f"""You are {config.display_name}, a helpful assistant.

{config.description}

When you have finished helping the user, use the transfer_back tool to return control to the main agent.
"""


def _generate_gitignore() -> str:
    return """node_modules/
dist/
*.log
.DS_Store
"""
